File: org.summer.view.widget/src/org/summer/view/widget/input/RoutedUICommand.java
package org.summer.view.widget.input;

import org.summer.view.widget.ArgumentNullException;
import org.summer.view.widget.Type;

/// <summary> 
///     RoutedCommand with added UI Information.
/// </summary> 
//[TypeConverter("System.Windows.Input.CommandConverter, PresentationFramework, Version=" + BuildInfo.WCP_VERSION + ", Culture=neutral, PublicKeyToken=" + BuildInfo.WCP_PUBLIC_KEY_TOKEN + ", Custom=null")]
public class RoutedUICommand extends RoutedCommand
{
    /// <summary> 
    ///     Default Constructor - needed to allow markup creation
    /// </summary> 
    public RoutedUICommand() 
    {
    	super();
        _text = String.Empty; 
    }

    /// <summary>
    ///     Creates an instance of this class. 
    /// </summary>
    /// <param name="text">Descriptive and localizable text for the command</param> 
    /// <param name="name">Declared Name of the RoutedCommand for Serialization</param> 
    /// <param name="ownerType">Type that is registering the property</param>
    public RoutedUICommand(String text, String name, Type ownerType) 
        : this(text, name, ownerType, null)
    {
    }

    /// <summary>
    ///     Creates an instance of this class. 
    /// </summary> 
    /// <param name="text">Descriptive and localizable text for the command</param>
    /// <param name="name">Declared Name of the RoutedCommand for Serialization</param> 
    /// <param name="ownerType">Type that is registering the property</param>
    /// <param name="inputGestures">Default Input Gestures associated</param>
    public RoutedUICommand(String text, String name, Type ownerType, InputGestureCollection inputGestures)
        : base(name, ownerType, inputGestures) 
    {
        if (text == null) 
        { 
            throw new ArgumentNullException("text");
        } 
        _text = text;
    }

    /// <summary> 
    ///     Creates an instance of this class. Allows lazy initialization of InputGestureCollection and Text properties.
    /// </summary> 
    /// <param name="name">Declared Name of the RoutedCommand for Serialization</param> 
    /// <param name="ownerType">Type that is registering the property</param>
    /// <param name="commandId">An identifier assigned by the owning type to the command</param> 
    /*internal*/public RoutedUICommand(String name, Type ownerType, byte commandId):base(name, ownerType, commandId)
    {
    }

    /// <summary>
    ///     Descriptive and localizable text for the command. 
    /// </summary> 
    public String Text
    { 
        get
        {
            if(_text == null)
            { 
                _text = GetText();
            } 
            return _text; 
        }
        set 
        {
            if (value == null)
            {
                throw new ArgumentNullException("value"); 
            }
            _text = value; 
        } 
    }

    /// <summary>
    ///    Fetches the text by invoking the GetUIText function on the owning type.
    /// </summary>
    /// <returns>The text for the command</returns> 
    private String GetText()
    { 
        if(OwnerType == typeof(ApplicationCommands)) 
        {
            return ApplicationCommands.GetUIText(CommandId); 
        }
        else if(OwnerType == typeof(NavigationCommands))
        {
            return NavigationCommands.GetUIText(CommandId); 
        }
        else if(OwnerType == typeof(MediaCommands)) 
        { 
            return MediaCommands.GetUIText(CommandId);
        } 
        else if(OwnerType == typeof(ComponentCommands))
        {
            return ComponentCommands.GetUIText(CommandId);
        } 
        return null;
    } 

   private String _text;
}

File: summer/WebContent/js/system/windows/input/RoutedUICommand.js
/**
 * RoutedUICommand
 */

define(["dojo/_base/declare", "system/Type", "input/RoutedCommand"], 
		function(declare, Type, RoutedCommand){
	var RoutedUICommand = declare("RoutedUICommand", RoutedCommand,{
	 	"-chains-": {
	      constructor: "manual"
	    },
		constructor:function(text){
			if(text === undefined){
	            this._text = String.Empty; 
			}else{
				this._text = text;
			}
		},
		
	       /// <summary>
        ///    Fetches the text by invoking the GetUIText function on the owning type.
        /// </summary>
        /// <returns>The text for the command</returns> 
//        private string 
		GetText:function()
        { 
            if(OwnerType == typeof(ApplicationCommands)) 
            {
                return ApplicationCommands.GetUIText(CommandId); 
            }
            else if(OwnerType == typeof(NavigationCommands))
            {
                return NavigationCommands.GetUIText(CommandId); 
            }
            else if(OwnerType == typeof(MediaCommands)) 
            { 
                return MediaCommands.GetUIText(CommandId);
            } 
            else if(OwnerType == typeof(ComponentCommands))
            {
                return ComponentCommands.GetUIText(CommandId);
            } 
            return null;
        } 
	});
	
	Object.defineProperties(RoutedUICommand.prototype,{
		 /// <summary>
        ///     Descriptive and localizable text for the command. 
        /// </summary> 
//        public string 
        Text:
        { 
            get:function()
            {
                if(this._text == null)
                { 
                	this._text = GetText();
                } 
                return _text; 
            },
            set:function(value) 
            {
                if (value == null)
                {
                    throw new ArgumentNullException("value"); 
                }
                this._text = value; 
            } 
        }  
	});
	
	Object.defineProperties(RoutedUICommand,{
		  
	});
	
	RoutedUICommand.Build = function(){
		var obj = new RoutedUICommand();
		RoutedCommand.prototype.constructor.call(obj, null, null, null);
		return obj;
	};
	
	RoutedUICommand.BuildSST = function(/*string*/ text, /*string*/ name, /*Type*/ ownerType){
		var obj = new RoutedUICommand(text);
		RoutedCommand.prototype.constructor.call(obj, name, ownerType, null);
		return obj;
	};
	
	RoutedUICommand.BuildSTC = function(/*string*/ name, /*Type*/ ownerType, /*byte*/ commandId){
		var obj = new RoutedUICommand(null);
		RoutedCommand.prototype.constructor.call(obj, name, ownerType, commandId);
		return obj;
	};
	
	RoutedUICommand.BuildSSTI = function(/*string*/ text, /*string*/ name, /*Type*/ ownerType, /*InputGestureCollection*/ inputGestures){
		var obj = new RoutedUICommand(text);
		RoutedCommand.prototype.constructor.call(obj, name, ownerType, inputGestures);
		return obj;
	};
	
	RoutedUICommand.Type = new Type("RoutedUICommand", RoutedUICommand, [RoutedCommand.Type]);
	return RoutedUICommand;
});


